### PR TITLE
4.1 Lessons about Refactoring and Evolution

### LINKED ISSUE #220: 4.1 - Lessons Evolution & Refactoring

### PR DESCRIPTION
**Pull request by Duckapple**
Closes #220

### CODE REVIEW

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 1]
```diff
@@ -1 +1,97 @@
...
+
+Implementation of a proper test suite came later down the line, with
+The team was always aware of the fact that there should be more tests, and were even reminded when things broke in production (like when the register page broke completely).
+Though it is a lesson that the team has already learned in previous projects, here the team learned it once again.
+
+### Evolution
+
+Once the necessary refactoring steps were done, the team was capable of expanding the original implementation.
+
+#### Automation is king
```
Comment (Jakobis) [inline report/41_Lessons-Evolution-&-Refactoring.md]: Maybe a sligthly too unacademic title for a report

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 2]
```diff
@@ -1 +1,97 @@
...
+
+#### Make sure it all works
+
+In order to ensure that the refactor didn't break anything, tests could be useful.
+Only problem is that the previous tests were an API test script along with Python unit tests.
+
+In order to see immediate results, a Github Action was created to automate running the API test.
+That didn't really work, as it didn't catch all edge cases, so instead the project got configured to be tested with the API simpulator.
+
+Implementation of a proper test suite came later down the line, with
```
Comment (Jakobis) [inline report/41_Lessons-Evolution-&-Refactoring.md]: With what?

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 3]
```diff
@@ -1 +1,97 @@
...
+#### Graphs are motivational
+
+After the simulator had been live for a few weeks, the statistics showed that this project lacked behind the other participating projects by quite a big margin.
+Having this as a motivator to improve was vital to finding some of the bugs in the system.
+
+First, it was discovered that the project lacked behind because response times were highly influential for the graph trend lines.
+It was then considered whether HTTPS and Traefik routing were the thing slowing down the simulator, but removing those things only made very slight improvements, so they were re-enabled.
+The actual reason turned out to be the server location, as will be described in [TODO: 4.2 Operations](42_Lessons-Operation.md).
+
+Without the graph as a motivator, these things would never have been discovered.
```
Comment (Jakobis) [inline report/41_Lessons-Evolution-&-Refactoring.md]: Arguably, this sections is not so much about graphs as it is about statistics, or likely, about competition. An interesting reflection would be how to recreate the feeling of competition in a real life environment, since this situation with 10 groups doing the same project would never happen in real life.

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 4]
```diff
@@ -1 +1,97 @@
...
+
+#### Proper communication
+
+A problem that the team never got to improve on was communication, especially of the asynchronous kind.
+When co-located, the team managed to get lots of work done and reviewed, but at the times between co-located sessions and after the pandemic hit the world, then reviews were slow and impractical.
```
Comment (Jakobis) [inline report/41_Lessons-Evolution-&-Refactoring.md]: This sentence is pretty weird, although it might be because of the "then" in "then reviews were..."

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 5]
```diff
@@ -1 +1,97 @@
...
+
+A possible reason could be that reminding people to review is easier and more direct when working next to each other.
+Another could be that the other courses and work got in the way.
+
+A third possibility could be that there were no channels for politely nudging towards a review.
+The team tried to alleviate this with the `#review` channel on Discord, where anyone could push for a review.
+This helped in some ways, but simply pushed the notification into its own little box that was easier to forget or ignore.
```
Comment (Jakobis) [inline report/41_Lessons-Evolution-&-Refactoring.md]: I disagree. The review channel meant I got way better at looking at reviews, because I was more likely to check discord than github, and Discord is clearer about notification source.

**Review by Jakobis — CHANGES_REQUESTED** [review]

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 6]
```diff
@@ -1 +1,97 @@
...
+
+The first task in this project was to refactor the provided system twice; first to Python 3, then a tech stack of the team's own choice.
+
+#### Python 2 to Python 3
+
+The conversion from Python 2 to 3 was almost symbolic for easy wins in DevOps.
+To do the conversion, a tool called [2to3](http://python3porting.com/2to3.html) was used; feed Python 2 in, get Python 3 out.
+
+It worked almost completely right away, which was surprising as the introduction to the tool in class said to be careful that nothing broke.
+Only thing needing a fix was the tests, which were previously dependent on the more raw string handling of Python 2, but that was not terribly hard to fix (See [This PR](https://github.com/jlndk/devoops/pull/6) for more).
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: It's  actually Unicode handling in python 2 vs 3

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 7]
```diff
@@ -1 +1,97 @@
...
+
+#### Python 3 with SQLite to ASP.NET with PostgreSQL
+
+The following conversion was much larger than the previous.
+The team decided to refactor the web server to an ASP.NET Core 3 application with PostgreSQL as the DBMS, as the language was familiar to some members, while being something stable to learn for the rest of the team.
+
+This refactor turned out to be quite monolithic, in part due to everything being changed at once.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: remove " in part "

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 8]
```diff
@@ -1 +1,97 @@
...
+Learning the syntax of Razor and Flask at the same time provided somewhat of a challenge, and porting to a MVC architecture at the same time didn't make this transition easier.
+
+The arguably biggest part was the implementation of Entity Framework (EF) and porting to PostgreSQL.
+For each part of the model, an entity would have to be established in code, and a context and repository layer for accessing and changing that entity would have to be written.
+Then, on top of that, an API layer was necessary to access and mutate this state.
+On the bright side, this completely eliminated plain text SQL queries in the code itself.
+
+If such a large refactor was to be done some other time, then a more incremental approach might be better.
+If the database was first changed to rely on another DBMS, and _then_ changed to work with EF, then the refactor could become more incremental and thus less monolithic.
+The only problem here is that it would equate to doing double the work, and when a refactor has to be done in a small timeframe, then double work doesn't seem practical.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md:33]: Maybe add that because we did all this work we didnt have any the next work. And it also made every change later easier.

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 9]
```diff
@@ -1 +1,97 @@
...
+### Evolution
+
+Once the necessary refactoring steps were done, the team was capable of expanding the original implementation.
+
+#### Automation is king
+
+Early on in the development process, not many automatic steps were set up in the project pipeline.
+This made evolution much harder than it had to be, as tests would have to be manually run all the time to verify builds, and releases would have to be done manually.
+
+In the pipeline, the team has focused on having many actions done automatically, as mentioned in [TODO: 2.3: CI/CD](23_Process-CI-CD-Description.md).
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: Fix Todo.

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 10]
```diff
@@ -1 +1,97 @@
...
+The arguably biggest part was the implementation of Entity Framework (EF) and porting to PostgreSQL.
+For each part of the model, an entity would have to be established in code, and a context and repository layer for accessing and changing that entity would have to be written.
+Then, on top of that, an API layer was necessary to access and mutate this state.
+On the bright side, this completely eliminated plain text SQL queries in the code itself.
+
+If such a large refactor was to be done some other time, then a more incremental approach might be better.
+If the database was first changed to rely on another DBMS, and _then_ changed to work with EF, then the refactor could become more incremental and thus less monolithic.
+The only problem here is that it would equate to doing double the work, and when a refactor has to be done in a small timeframe, then double work doesn't seem practical.
+
+#### Make sure it all works
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: Better title maybe Post refactor Verification

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 11]
```diff
@@ -1 +1,97 @@
...
+
+In order to ensure that the refactor didn't break anything, tests could be useful.
+Only problem is that the previous tests were an API test script along with Python unit tests.
+
+In order to see immediate results, a Github Action was created to automate running the API test.
+That didn't really work, as it didn't catch all edge cases, so instead the project got configured to be tested with the API simpulator.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: > In order to see immediate results, a Github Action was created to automate running the API test.
> That didn't really work, as it didn't catch all edge cases, so instead the project got configured to be tested with the API simpulator.

What does these two senteces try to say? It cant make sense of them.
Also "simpulator"

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 12]
```diff
@@ -1 +1,97 @@
...
+
+Once the necessary refactoring steps were done, the team was capable of expanding the original implementation.
+
+#### Automation is king
+
+Early on in the development process, not many automatic steps were set up in the project pipeline.
+This made evolution much harder than it had to be, as tests would have to be manually run all the time to verify builds, and releases would have to be done manually.
+
+In the pipeline, the team has focused on having many actions done automatically, as mentioned in [TODO: 2.3: CI/CD](23_Process-CI-CD-Description.md).
+This was, at first, because the course demanded it, but once the team started using automatic steps, then the value was apparent.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: Using the team all over the place seems quite unnatural and just a copy paste replacement for "we" or "our". Rather you reword the sentence to have it not be there.

File: report/41_Lessons-Evolution-&-Refactoring.md (hunk 5 above)
Comment (Duckapple) [reply report/41_Lessons-Evolution-&-Refactoring.md]: Please see whether you agree with the update.

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 13]
```diff
@@ -1 +1,97 @@
...
+The pipeline was expanded repeatedly, with the steps themselves even being expanded, as is the case with `e2e.yml`.
+
+The takeaway here is that automation is very useful, as these manual, menial tasks could now be carried out automatically.
+A focus for the team members moving forward is to set up robust automation steps in future projects as soon as possible.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: Maybe reword

> A focus for the team members moving forward is to set up robust automation steps in future projects as soon as possible.

To be more a clear that it was something we learned.

Something like:

`The takeaway here is that automation is very useful, as these manual, menial tasks could now be carried out automatically.
It is therefore now a focus for the team members moving forward is to set up robust automation steps in future projects as soon as possible. As it would remove the need for doing menial tasks`

**Review by Hjaltesorgenfrei — CHANGES_REQUESTED** [review]

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 14]
```diff
@@ -1 +1,101 @@
...
+
+#### Choose scalable technologies early
+
+The team knew that choosing scalable techonologies early was important.
+If a choice to switch from a monolithic server to a distributed system comes too late, then the changes that will have to be made could become quite big, resulting in a lot of technical debt.
+
+This was why dockerization came early for the project, with separation between components for easier clusterization.
+The team didn't quite get Docker Swarm to work, but because these considerations were done early, Docker Swarm was actually obtainable, instead of being buried in debt from not having a RESTful API, for example.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: Not totally happy with this last sentence

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 15]
```diff
@@ -1 +1,101 @@
...
+
+After the simulator had been live for a few weeks, the statistics showed that this project lacked behind the other participating projects by quite a big margin.
+Having this as a motivator to improve was vital to finding some of the bugs in the system.
+
+First, it was discovered that the project lacked behind because response times were highly influential for the graph trend lines.
+It was then considered whether HTTPS and Traefik routing were the thing slowing down the simulator, but removing those things only made very slight improvements, so they were re-enabled.
+The actual reason turned out to be the server location, as will be described in [TODO: 4.2 Operations](42_Lessons-Operation.md).
+
+Without the graph, the statistics, and the competitive nature of the team as motivators, these things would never have been discovered.
+Replication of this feeling would be difficult, as there rarely are 10 teams competing to implement the best and fastest project when looking at the job market.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: But you do have competitors. So having their statictics up, might work instead.

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 16]
```diff
@@ -1 +1,101 @@
...
+Gamification of collected metrics might be able to lead teams to improve these things, for example through having a big screen with several desired trend lines/thresholds along with actual statistics.
+The more prominent the gamification, the more this competitive nature could drive the team to do better.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: I really like this part. Nice.

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 17]
```diff
@@ -1 +1,101 @@
...
+At the times between co-located sessions and after the pandemic hit the world, the reviews were slow and impractical.
+
+A possible reason could be that reminding people to review is easier and more direct when working next to each other.
+Another could be that the other courses and work got in the way.
+
+A third possibility could be that there were no channels for politely nudging towards a review.
+The team tried to alleviate this with the `#review` channel on Discord, where anyone could push for a review.
+This helped in some ways, as Discord is a more prominently used communication tool among the developers.
+
+Though the team hasn't found better solutions than just being more demanding towards each other, it acknowledges that this is a problem that needs improving somehow down the line.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: I wrote about something similiar in 4.3, it might be worth making them fitter better together.

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 18]
```diff
@@ -1 +1,101 @@
...
+
+This doesn't mean that all communication was bad, far from it.
+Discord and GitHub were great places for collecting information, the problems were mainly with the project management idea of assigning reviewers and such.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: > project management idea of assigning reviewers and such.

is a awkward sentence.

**Review by Hjaltesorgenfrei — CHANGES_REQUESTED** [review]

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 19]
```diff
@@ -66,7 +66,8 @@ The team knew that choosing scalable techonologies early was important.
 If a choice to switch from a monolithic server to a distributed system comes too late, then the changes that will have to be made could become quite big, resulting in a lot of technical debt.
 
 This was why dockerization came early for the project, with separation between components for easier clusterization.
-The team didn't quite get Docker Swarm to work, but because these considerations were done early, Docker Swarm was actually obtainable, instead of being buried in debt from not having a RESTful API, for example.
+Because the considerations were done early, Docker Swarm was actually obtainable, instead of being buried in debt from not having a RESTful API, for example.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: "was actually obtainable"
It would still have been, had we done it differently. It was just easier.

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 20]
```diff
@@ -66,7 +66,8 @@ The team knew that choosing scalable techonologies early was important.
...
+The team unfortunately didn't get Docker Swarm to work though the work put into trying can be found on the `swarm` branch.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: Using "The Team" makes for a awkward sentence. 
The problem with "we" is not the word, but the way it structures the sentence.

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 21]
```diff
@@ -78,8 +79,8 @@ It was then considered whether HTTPS and Traefik routing were the thing slowing
 The actual reason turned out to be the server location, as will be described in [TODO: 4.2 Operations](42_Lessons-Operation.md).
 
 Without the graph, the statistics, and the competitive nature of the team as motivators, these things would never have been discovered.
-Replication of this feeling would be difficult, as there rarely are 10 teams competing to implement the best and fastest project when looking at the job market.
-Gamification of collected metrics might be able to lead teams to improve these things, for example through having a big screen with several desired trend lines/thresholds along with actual statistics.
+Replication of this feeling later could be done by collecting metrics from competetitors, if there are any, and gamifying those metrics.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: > Replication of this feeling later could be done by collecting metrics from competetitors, if there are any, and gamifying those metrics.

Change to

"Replication of this feeling could be done by collecting metrics from competitors, and gamifying these."

File: report/41_Lessons-Evolution-&-Refactoring.md [hunk 22]
```diff
@@ -78,8 +79,8 @@ It was then considered whether HTTPS and Traefik routing were the thing slowing
...
+This might be able to lead teams to improve those metrics, for example through having a big screen with several desired trend lines/thresholds along with actual statistics.
```
Comment (Hjaltesorgenfrei) [inline report/41_Lessons-Evolution-&-Refactoring.md]: This sentence is awkward.

**Review by Hjaltesorgenfrei — CHANGES_REQUESTED** [review]

File: report/41_Lessons-Evolution-&-Refactoring.md (hunk 12 above)
Comment (Duckapple) [reply report/41_Lessons-Evolution-&-Refactoring.md]: Trimmed from 23 to 3, if you can eliminate the rest the go ahead

**Review by Jakobis — APPROVED** [review]

File: report/41_Lessons-Evolution-&-Refactoring.md (hunk 12 above)
Comment (Hjaltesorgenfrei) [reply report/41_Lessons-Evolution-&-Refactoring.md]: The last are fine.

**Review by Hjaltesorgenfrei — APPROVED** [review]
